glean-core/src/scheduler.rs

```diff
@@ -0,0 +1,164 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! # Metrics Ping Scheduler
+//!
+//! The Metrics Ping Scheduler (MPS) is responsible for scheduling "metrics" pings.
+//! It implements the spec described in
+//! [the docs](https://mozilla.github.io/glean/book/user/pings/metrics.html#scheduling)
+
+use crate::metrics::{DatetimeMetric, StringMetric, TimeUnit};
+use crate::{CommonMetricData, Glean, Lifetime, INTERNAL_STORAGE};
+use chrono::prelude::*;
+use once_cell::sync::OnceCell;
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+
+const SCHEDULED_HOUR: u32 = 4;
+
+static THREAD_GENERATION: OnceCell<Arc<AtomicU32>> = OnceCell::new();
+
+/// Performs startup checks to decide when to schedule the next "metrics" ping collection.
+/// Must be called before draining the preinit queue.
+/// (We're at the Language Bindings' mercy for that)
+pub fn schedule(glean: &Glean) {
+    let now = Local::now();
+
+    schedule_internal(&glean, now)
+}
+
+fn schedule_internal(glean: &Glean, now: DateTime<Local>) {
+    let last_sent_build_metric = get_last_sent_build_metric();
+    if let Some(last_sent_build) = last_sent_build_metric.get_value(&glean, INTERNAL_STORAGE) {
+        // XXX: If `app_build` is longer than StringMetric's max length, we will always
```

Comment:
Hm, that's a question: If the length is too long that it gets truncated, will the error we record (`invalid_overflow` methinks) show up in a place we'd see it? Or will it be hidden because I'm using `INTERNAL_STORAGE`?

Because if we report it through error reporting, then Woo to the Hoo, job done.